hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/transaction/IngestionPrimaryWriterBasedConflictResolutionStrategy.java

```diff
@@ -0,0 +1,116 @@
+   */
+  @Override
+  public Stream<HoodieInstant> getCandidateInstants(HoodieTableMetaClient metaClient, HoodieInstant currentInstant,
+                                                    Option<HoodieInstant> lastSuccessfulInstant) {
+    HoodieActiveTimeline activeTimeline = metaClient.reloadActiveTimeline();
+    if ((REPLACE_COMMIT_ACTION.equals(currentInstant.getAction())
+          && ClusteringUtils.isClusteringCommit(metaClient, currentInstant))
+        || COMPACTION_ACTION.equals(currentInstant.getAction())) {
+      return getCandidateInstantsForTableServicesCommits(activeTimeline, currentInstant);
+    } else {
+      return getCandidateInstantsForNonTableServicesCommits(activeTimeline, currentInstant);
+    }
+  }
+
+  private Stream<HoodieInstant> getCandidateInstantsForNonTableServicesCommits(HoodieActiveTimeline activeTimeline, HoodieInstant currentInstant) {
+
+    // To findout which instants are conflicting, we apply the following logic
+    // Get all the completed instants timeline only for commits that have happened
+    // since the last successful write based on the transition times.
+    // We need to check for write conflicts since they may have mutated the same files
+    // that are being newly created by the current write.
+    List<HoodieInstant> completedCommitsInstants = activeTimeline
+        .getTimelineOfActions(CollectionUtils.createSet(COMMIT_ACTION, REPLACE_COMMIT_ACTION, COMPACTION_ACTION, DELTA_COMMIT_ACTION))
+        .filterCompletedInstants()
+        .findInstantsModifiedAfterByStateTransitionTime(currentInstant.getTimestamp())
+        .getInstantsOrderedByStateTransitionTime()
+        .collect(Collectors.toList());
+    LOG.info(String.format("Instants that may have conflict with %s are %s", currentInstant, completedCommitsInstants));
+    return completedCommitsInstants.stream();
+  }
+
+  /**
+   * To find which instants are conflicting, we apply the following logic
+   * Get both completed instants and ingestion inflight commits that have happened since the last successful write.
+   * We need to check for write conflicts since they may have mutated the same files
+   * that are being newly created by the current write.
+   */
+  private Stream<HoodieInstant> getCandidateInstantsForTableServicesCommits(HoodieActiveTimeline activeTimeline, HoodieInstant currentInstant) {
+    // Fetch list of completed commits.
+    Stream<HoodieInstant> completedCommitsStream =
+        activeTimeline
+            .getTimelineOfActions(CollectionUtils.createSet(COMMIT_ACTION, REPLACE_COMMIT_ACTION, COMPACTION_ACTION, DELTA_COMMIT_ACTION))
+            .filterCompletedInstants()
+            .findInstantsModifiedAfterByStateTransitionTime(currentInstant.getTimestamp())
+            .getInstantsAsStream();
+
+    // Fetch list of ingestion inflight commits.
+    Stream<HoodieInstant> inflightIngestionCommitsStream =
```

Comment:
One reminder, not all inflight commits have metadata, for example, bulkInsert's inflight metadata is empty,
In this case, IngestionPrimaryWriterBasedConflictResolutionStrategy will be invalid

Reply:
@Zouxxyy I idea behind creating IngestionPrimaryWriterBasedConflictResolutionStrategy, is to give preference for Ingestion writers over clustering writers. 
For example, you have clustering plans as mutable plans which can be created and removed by rollback, this is achievable by using SparkAllowUpdateStrategy. Clustering writers and Ingestion writers can run simultaneously on the same partition. If clustering writer were to complete writing first and about to create a completed replacecommit, in that stage it will check if there are any pending inflight ingestion instants in the timeline and fail if it finds an overlap.
For bulk insert operations, they create new files so there wont be ay conflict with clustering writers, so in that case replacecommit is allowed to commit. There could be a case where if users are using insert API, but it can still write to existing files, if small file handling is enabled. Even then possibility of overlap between clustering and ingestion can be reduced and even if there is overlap you can fail clustering commit and allow Ingestion to commit even if it is arriving late.

Reply:
> in that stage it will check if there are any pending inflight ingestion instants in the timeline and fail if it finds an overlap.

Thanks for the reply, what I mean is now we use the metadata of pending ingestion instant for conflict detection here, but not all inflight commits have metadata, e.g. the .inflght file written by flink is empty.

And can you help review this https://github.com/apache/hudi/pull/9220 which improve the current stragies, thanks

Reply:
Ok. I am not much familiar with Flink, but on the Spark engine the inflight commits that are doing upsert operations should have partition to fileIds(that needs to be modified) present in the .inflight file, this is actually done after the WorkloadProfile object is loaded. So, conflict resolution will only see those file ids that are getting modified not new ones. During conflict resolution that is all we care, since there won't be any overlap between new files. Sure, I will review the RFC.